### ISSUE TITLE
Document setup and workflow for multi-user editing

### ISSUE THREAD

**Issue by maptastik** [koordinates/kart#614]
# Problem
I have been testing Kart for a few days and am quite impressed by the promise it holds as an alternative to ArcSDE for versioned, multi-user editing of spatial data. That Kart is built on git (as I understand it) suggests to me that multi-user editing should be possible, but I have been unable to successfully setup a project where that is possible.

# Suggested Solution

It would be great to have an additional guide about using kart for multi-user editing. I acknowledge that I might be missing something obvious in the documentation or that I have a fundamental misunderstanding of Kart's purpose, but it is not clear to me how one would be able to have multiple users working on the same dataset. I suspect there are many folks like me who have needs for multi-user, versioned spatial data editing, but for whom ArcSDE seems to be overkill or is otherwise not an option. A basic how-to guide on multi-user editing might make Kart seem a bit more useful to folks trying to escape the SDE trap.

# What I've done

## Worked through documentation

I have spent quite a bit of time in the documentation for Kart with much focus on the [Quick Guide](https://docs.kartproject.org/en/latest/pages/quick_guide.html) and [Basic Usage Tutorial](https://docs.kartproject.org/en/latest/pages/basic_usage_tutorial.html). With these guides, I have been able to successfully, create a Kart repository, import a working dataset, edit the working data, commit changes, revert changes, edit on a branch, and merge changes from one branch into another.

## Watched *Spatial data versioning with the Kart QGIS Plugin with Victor Olaya*

I have followed along with [Victor Olaya's presentation](https://www.youtube.com/watch?v=aABc3JrgJUY) on the Kart QGIS Plugin. I know this is not about Kart alone, but I did find that Victor spoke much about the fundamentals of Kart that are made accessible via the QGIS plugin's GUI. I believe at one point Victor indicates that multi-user editing is possible, but his presentation was focused on the plugin functionality and thus used a simple, single-user workflow.

## Create repo with working data included, then clone

Not finding explicit documentation on a multi-user approach, I tried emulating the familiar workflow I've used when in multi-user editing scenarios on GitHub. I used the Basic Usage Tutorial as a guide as well. The process was as follows:

1. Created a new Kart repository (`buildings-project`) and imported a dataset (`wellington-building-outlines.gpkg`). 
2. I cloned the Kart repository to a new directory using `kart clone` (`buildings-project-clone-1`)
3. In QGIS, I loaded the working dataset from the clone, `buildings-project-clone-1` and edited the geometry of a feature.
4. I saved the edit and committed the change (`kart commit -m "update message"`).
5. I attempted to push the change from the clone, `buildings-project-clone-`, to the repository created in step 1, `buildings-project`, by running `kart push`.

At this point I got the following log and error messages:
```
Enumerating objects: 10, done.
Counting objects: 100% (10/10), done.
Writing objects: 100% (10/10), 4.38 KiB | 4.38 MiB/s, done.
Total 10 (delta 0), reused 10 (delta 0), pack-reused 0
remote: error: refusing to update checked out branch: refs/heads/main
remote: error: By default, updating the current branch in a non-bare repository
remote: is denied, because it will make the index and work tree inconsistent
remote: with what you pushed, and will require 'git reset --hard' to match
remote: the work tree to HEAD.
remote:
remote: You can set the 'receive.denyCurrentBranch' configuration variable
remote: to 'ignore' or 'warn' in the remote repository to allow pushing into
remote: its current branch; however, this is not recommended unless you
remote: arranged to update its work tree to match what you pushed in some
remote: other way.
remote:
remote: To squelch this message and still keep the default behaviour, set
remote: 'receive.denyCurrentBranch' configuration variable to 'refuse'.
To C:/path/to/buildings-project
 ! [remote rejected]     main -> main (branch is currently checked out)
error: failed to push some refs to 'C:/path/to/buildings-project'
```

This is [the workflow that I shared out to Twitter a few days ago](https://twitter.com/maptastik/status/1518613302397480961). I received [an explanation](https://twitter.com/KartForData/status/1519122860954243072) from someone running the Kart twitter handle about why this did not work. As I understand it, the remote, `buildings-project`, is not expected to have any working data in it. 

This challenged my expectations of how a multi-user workflow would work. I was expecting to have a dataset in the remote, clone it into another directory or machine and then push and pull changes to and from the remote, respectively. This *seems* to be how working with repo on, for instance, GitHub works. I emphasize "seems" because I acknowledge I am definitely misunderstanding something about what is actually going on when cloning and pushing and pulling changes from GitHub works. I also recognize that Kart is not GitHub and that I shouldn't expect assumptions from working with GitHub to necessarily carry over directly to Kart.

## Create repo with no working data, and clone

This workflow follows my understanding of the advice from the Kart twitter account.

1. Created a new Kart repository (`buildings-project`). 
2. I cloned the Kart repository to a new directory using `kart clone building-project building-project-bare-clone-1` (`buildings-project-bare-clone-1`)
3. Attempted to imported a dataset (`wellington-building-outlines.gpkg`) into the cloned repo, `buildings-project-bare-clone-1` by running `kart import ../wellington-building-outlines.gpkg`.

At this point I got stuck. The command ran with the following messages:

```
Starting git-fast-import...
Importing 75,409 features from wellington-building-outlines.gpkg:wellington_building_outlines to wellington_building_outlines/ ...
Added 75,409 Features to index in 9.4s
Overall rate: 8044 features/s)
Closed in 0s
Joining 8 parallel-imported trees...
Joined trees in 0s
```

Although it appears something happened successfully, there was not a working copy created in `buildings-project-bare-clone-1`. In the spirit "just try something and see what happens," I ran the import command again and got the following error:

```
Error: Dataset path 'wellington_building_outlines' conflicts with existing path 'wellington_building_outlines'
```

So something did indeed happen, but it's unclear to me what that was.

# TLDR

- Kart seems like it should be able to handle multi-user editing
- It's not clear to me how to setup Kart for multi-user editing
- A user-guide on multi-user editing would be helpful

Thanks for your consideration of this matter and I'll be happy to provide additional information that is certainly missing here.

**Comment by hamishcampbell** [comment]
Hi,

It's important to understand the difference between a bare repository and a normal repository with a working copy.

  * A normal repository (as created by default on `kart clone`) contains the entire repository contents in the `.kart` folder and maintains a working copy checked out into the location specified, e.g. a GeoPackage on disk. This is where users make changes to the working copy data, commit changes, and run `push` and `pull` commands.
  * A bare repository only contains the repository contents (the contents of the `.kart` folder) and does not include a working copy. A remote that can be cloned, and receives changes via `push` will almost always be a bare repository, and could be hosted on a service like GitHub or on your file-system.

### Creating a bare Repository

You can create a new, empty Kart repository using `kart init --bare`. For example

```
~$ mkdir buildings
~$ cd buildings
~/buildings/$ kart init --bare
```

And import data via `kart import`:

~/buildings/$ kart import ../wellington-building-outlines.gpkg
```

It looks like this is the process you followed above. The end result is a Kart repository that does not have a working copy GPKG. This is expected – this repository is expected to receive changes to the repository data from other clones, rather than provide a working space for editing and committing changes.

You can verify that the data exists in the repository with `kart data ls`:

```
~/buildings/$ kart data ls
wellington_building_outlines
```

The error when you try to import the data again is expected, as the data is already present.

### Cloning the repository with a working copy

In a separate location you can now clone from the first location:

```
~$ kart clone ~/buildings/ buildings-new
~$ cd buildings-new
~/buildings-new/$ ls
KART_README.txt    buildings-new.gpkg
```

This new clone is a repository that includes the working copy checked out as a GeoPackage.

### Committing and pushing changes

You can make changes to the data, commit changes, and push those chants back to the first bare repository:

```
~/buildings-new/$ kart commit -m "commit my edits"
~/buildings-new/$ kart push

Enumerating objects: 10, done.
Counting objects: 100% (10/10), done.
Writing objects: 100% (10/10), 2.30 KiB | 2.30 MiB/s, done.
Total 10 (delta 0), reused 10 (delta 0), pack-reused 0
To ~/buildings/
   fbcb08074..e49405c59  main -> main
```

Hopefully this helps clarify how multi-user editing using a file system repository works. We will try to incorporate this into a tutorial to help other people testing the same way.

**Comment by maptastik** [comment]
Thank you @hamishcampbell for taking the time to outline this process and helping me understand why my original attempts had failed. Your explanation for how to clone, commit, and push were all very helpful as well.
